keps/sig-multicluster/5339-clusterprofile-plugin-credentials/README.md

```diff
@@ -0,0 +1,619 @@
+  // If empty, system roots should be used.
+  // +listType=atomic
+  // +optional
+  CertificateAuthorityData []byte `json:"certificate-authority-data,omitempty"`
+  // ProxyURL is the URL to the proxy to be used for all requests to this
+  // cluster.
+  // +optional
+  ProxyURL string `json:"proxy-url,omitempty"`
+  // Config holds additional config data that is specific to the exec
+  // plugin with regards to the cluster being authenticated to.
+  //
+  // This data is sourced from the clientcmd Cluster object's
+  // extensions[client.authentication.k8s.io/exec] field:
+  //
+  // clusters:
+  // - name: my-cluster
+  //   cluster:
+  //     ...
+  //     extensions:
+  //     - name: client.authentication.k8s.io/exec  # reserved extension name for per cluster exec config
+  //       extension:
+  //         audience: 06e3fbd18de8  # arbitrary config
+  //
+  // In some environments, the user config may be exactly the same across many clusters
+  // (i.e. call this exec plugin) minus some details that are specific to each cluster
+  // such as the audience.  This field allows the per cluster config to be directly
+  // specified with the cluster info.  Using this field to store secret data is not
+  // recommended as one of the prime benefits of exec plugins is that no secrets need
+  // to be stored directly in the kubeconfig.
+  // +optional
+  Config runtime.RawExtension `json:"config,omitempty"`
+}
+```
+
+This data must be created from the ClusterProfile, as described in the following section.
+
+### Standardizing the Endpoint Property
+
+In order to populate the Cluster object that the exec provider requires, we standardize the following properties.
+
+Property: `multicluster.sigs.io/execprovider-cluster`
+Value: json containing the Cluster struct (see above).
+Example Value: "{\"Server\":\"https://example.com:443\"}"
+
+### Configuring plugins in the controller
+
+Plugins are designated by a hardcoded string, for example, "gke" for GKE Clusters, which allows the controller
+to attach a different binary name or path for the the binary. It is expected that the library will have a mapping from
+its supported plugin names to the expected binary to call. It is expected that the mapping is provided via a flag on the controller.
+
+```
+./controller ... --clusterprofile-creds-providers "gke=gke-gcloud-auth-plugin"
+```
+
+### Selecting the right plugin
+
+The plugin is designated by the clusterprofile, which designates the type of provider that is expected to be called.
+The clusterprofile has a property to designate the plugin.
+
+Property: `multicluster.sigs.io/execprovider-name`
+Value: string that is expected from the controller to satisfy via plugins.
+Example Value: "gke"
+
+### External credentials Provider plugin mechanism
+
+In order to call the plugin, the library execs the plugin defined in the configuration. It passes the Cluster information that
+was obtained from the ClusterProfile.
```

Comment:
this is where we need an extra field/status in the ClusterProfile.  We can also make this as a standard property if a field feels a bit too rigid.

Reply:
moved to a field.